pkg/rhcos/updatepayload/updatepayload.go

```diff
@@ -0,0 +1,157 @@
+// Package updatepayload extracts the RHCOS build version from an
+// OpenShift update payload.
+package updatepayload
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"io"
+	"io/ioutil"
+
+	"github.com/containers/image/docker"
+	"github.com/containers/image/image"
+	"github.com/containers/image/pkg/blobinfocache"
+	"github.com/containers/image/transports"
+	"github.com/containers/image/types"
+	"github.com/docker/distribution/reference"
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+func pull(ctx context.Context, pullSpec string) (types.ImageSource, types.Image, error) {
+	named, err := reference.ParseNamed(pullSpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	reference, err := docker.NewReference(named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sys := &types.SystemContext{}
+	imageSource, err := reference.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() {
+		if err != nil {
```

Comment:
this is confusing.. which err are we trying to check...?

Reply:
> this is confusing.. which err are we trying to check...?

If `FromUnparsedImage` (or the coming signature validation) fails, I want to be sure we close `imageSource`.  So this closure looks at the `err` that will be returned by `pull`, and, if it's not `nil`, closes `imageSource` for us.  That way `pull` callers can just return the error without having to worry about checking `imageSource` for `nil`-ness and possibly closing it themselves; they can just return pull's `err`.